smolyay/surrogate.py

```diff
@@ -361,7 +361,16 @@ def _mapdomain(x, old, new):
         if old.ndim != 2 or old.shape[1] != 2:
             raise TypeError('Domain should be a dim x 2 array')
 
-        return new[:,0]+(new[:,1]-new[:,0])*((x-old[:,0])/(old[:,1]-old[:,0]))
+        new_x = new[:,0]+(new[:,1]-new[:,0])*((x-old[:,0])/(old[:,1]-old[:,0]))
+        # clamp lower bound
+        flags = numpy.array(new_x < new[:, 0])
+        test = numpy.broadcast_to(new[:, 0],flags.shape)
+        new_x[flags] = test[flags]
+        # clamp upper bound
+        flags = numpy.array(new_x > new[:, 1])
+        test = numpy.broadcast_to(new[:, 1],flags.shape)
+        new_x[flags] = test[flags]
```

Comment:
No, it gives a lot of index errors.

Reply:
What is the test case that fails? I just ran this on my desktop and it worked fine:
```
>>> x = [2.1, -6]
>>> old = np.array([[0, 2],[-5, 5]])
>>> new = np.array([[-1, 1], [-1, 1]])
>>> new_x = new[:,0]+(new[:,1]-new[:,0])*((x-old[:,0])/(old[:,1]-old[:,0]))
>>> new_x
array([ 1.1, -1.2])
>>> flags = new_x < new[:, 0]
>>> new_x[flags] = new[flags, 0]
>>> flags = new_x > new[:, 1]
>>> new_x[flags] = new[flags, 1]
>>> new_x
array([ 1., -1.])
```

Reply:
When running the pytest, I got 28 errors.
FAILED tests/surrogate_test.py::test_initialization_1d - IndexError: boolean index did not match indexed array along dimension 0; di...
FAILED tests/surrogate_test.py::test_gradient_initialization_1d - IndexError: boolean index did not match indexed array along dimension 0; di...
FAILED tests/surrogate_test.py::test_initialization_2d - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_gradient_initialization_2d - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_map_function_1d - IndexError: boolean index did not match indexed array along dimension 0; di...
FAILED tests/surrogate_test.py::test_map_function_2d - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_surrogate_0[lu] - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_surrogate_0[lstsq] - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_surrogate_0[inv] - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_surrogate_1 - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_surrogate_0_shifted - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_surrogate_2[lu] - IndexError: boolean index did not match indexed array along dimension 0; di...
FAILED tests/surrogate_test.py::test_surrogate_2[lstsq] - IndexError: boolean index did not match indexed array along dimension 0; di...
FAILED tests/surrogate_test.py::test_surrogate_2[inv] - IndexError: boolean index did not match indexed array along dimension 0; di...
FAILED tests/surrogate_test.py::test_train_from_data - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_gradient_train_from_data - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_gradient_surrogate_3 - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...
FAILED tests/surrogate_test.py::test_gradient_surrogate_4 - IndexError: boolean index did not match indexed array along dimension 0; di...
FAILED tests/surrogate_test.py::test_error_solver - IndexError: boolean index did not match indexed array along dimension 0; di...
FAILED tests/surrogate_test.py::test_error_input_surrogate - IndexError: too many indices for array: array is 2-dimensional, but 3 were ...

Reply:
These are test failure messages, but it's impossible for me to figure out what is going wrong by you pasting a portion of the output when I also can't see your code.

You should work on a specific case that is failing, and you can ask for help on Slack (with the full error output) if you need it.

Reply:
The crux of the issue is that the shape of new_x and flags is the number of points you have x dimension while the shape of new is dimension x 2. In order to map booleans in flags onto new, you need to first broadcast new to the shape of new_x and flags. That's what the test variable does (which admittedly needs a better name).

Yours didn't error because all the shapes just happened to work out. new_x is (2, ), flags is (2, ), new is (2, 2), and new[:,0] is (2, ).

But in the pytests like in test_initialization_1d, new_x is (5, ), flags is (5, ), new is (1, 2), and new[:,0] is (1, ).

Reply:
Oh, yes, I see! In that case, I would recommend that you do a `for` loop over the dimensions (if there is more than one dimension). You don't want to use `broadcast_to` because it will do an allocation & copy every time this function gets called. There is also actually a function `clip` that would help with this. I think something like:
```
for i in range(self.dimension):
     numpy.clip(new_x[:, i], new[i, 0], new[i, 1], out=new_x[:, i])
```
You probably need to treat the 1d case specially since currently `x` is not required to be 2d.

Reply:
It works now.